Add Serverless section with AWS Lambda docs (#1974)
open-telemetry/opentelemetry.io: content/en/docs/instrumentation/js/serverless.md

---
title: Serverless
weight: 8
---

This guide will show you how to get started with tracing serverless functions using OpenTelemetry instrumentation libraries.

## AWS Lambda

Following the next steps will allow your AWS Lambda functions to use OpenTelemetry to send traces to the configured backend.

### Dependencies

Create an empty package.json:

```sh
npm init -y
```

Install dependencies used by the example.

```sh
npm install \
  @opentelemetry/api \
  @opentelemetry/auto-instrumentations-node \
  @opentelemetry/exporter-trace-otlp-http \
  @opentelemetry/instrumentation \
  @opentelemetry/sdk-trace-base \

>   @opentelemetry/sdk-trace-node

Reviewer: I think you don't need all of them because they have some interdependencies, can you verify what the smallest set of packages is, that you really need?

Author: Just verified that this is the smallest number of packages that I need

  instrumentations: [
    getNodeAutoInstrumentations({
      "@opentelemetry/instrumentation-aws-lambda": {
        disableAwsContextPropagation: true,
      },
    }),
  ],
});
```

Replace `<backend_url>` with the URL of your favorite backend to export all
traces to it. If you don't have one setup already, you can check out [Jaeger](https://www.jaegertracing.io/) or [Zipkin](https://zipkin.io/).

Note that `disableAwsContextPropagation` is set to true. The reason for this is
that the Lambda instrumentation tries to use the X-Ray context headers by
default, this results in a non-sampled context, which creates a
`NonRecordingSpan`.

More details can be found in the instrumentation
[documentation](https://www.npmjs.com/package/@opentelemetry/instrumentation-aws-lambda).

### AWS Lambda function handler

Now that you have a Lambda wrapper, you can create a simple handler that will
serve as a Lambda function. Save the following code as `handler.js`.

```javascript
/* handler.js */

"use strict";

const http = require("http");

module.exports.hello = async (event) => {
  http.get("https://opentelemetry.io/");

  return {
    statusCode: 200,
    body: "Success!",

>   };

Reviewer: this looks surprising, shouldn't there be some dependency between the http.get and the return of that function? like return the statusCode of the http.get in your return object?

  new OTLPTraceExporter(collectorOptions)
);

provider.addSpanProcessor(spanProcessor);
provider.register();

registerInstrumentations({
  instrumentations: [
    getNodeAutoInstrumentations({
      "@opentelemetry/instrumentation-aws-lambda": {
        disableAwsContextPropagation: true,
      },
    }),
  ],
});
```

Replace `<backend_url>` with the URL of your favorite backend to export all
traces to it. If you don't have one setup already, you can check out [Jaeger](https://www.jaegertracing.io/) or [Zipkin](https://zipkin.io/).

Note that `disableAwsContextPropagation` is set to true. The reason for this is
that the Lambda instrumentation tries to use the X-Ray context headers by
default, this results in a non-sampled context, which creates a
`NonRecordingSpan`.

More details can be found in the instrumentation
[documentation](https://www.npmjs.com/package/@opentelemetry/instrumentation-aws-lambda).

### AWS Lambda function handler

Now that you have a Lambda wrapper, you can create a simple handler that will
serve as a Lambda function. Save the following code as `handler.js`.

```javascript
/* handler.js */

"use strict";

const https = require('https');

> module.exports.hello = async (event) => {
>   https.get('https://opentelemetry.io/', res => {
>     return {
>       statusCode: res.statusCode,
>     };
>   }).on('error', err => {
>     console.log('Error: ', err.message);
>   });
> };

Reviewer: ```suggestion
module.exports.hello = async (event) => {
  return https.get('https://opentelemetry.io/', res => {
    return {
      statusCode: res.statusCode,
    };
  }).on('error', err => {
    console.log('Error: ', err.message);
  });
};
```
Not a lambda expert, but I assume you need to return something from the handler?

---
title: Serverless
weight: 8
---

This guide will show you how to get started with tracing serverless functions using OpenTelemetry instrumentation libraries.

## AWS Lambda

> Following the next steps will allow your AWS Lambda functions to use OpenTelemetry to send traces to the configured backend.

Reviewer 2: ```suggestion
The following show how to use OpenTelemetry to instrument AWS Lambda functions that send traces to a configured backend.
```

### Dependencies

> Create an empty package.json:

Reviewer 2: ```suggestion
First, creat an empty package.json:
```

```sh
npm init -y
```

> Install dependencies used by the example.

Reviewer 2: ```suggestion
Then install the required dependencies:
```

```sh
npm install \
  @opentelemetry/api \
  @opentelemetry/auto-instrumentations-node \
  @opentelemetry/exporter-trace-otlp-http \
  @opentelemetry/instrumentation \
  @opentelemetry/sdk-trace-base \
  @opentelemetry/sdk-trace-node
```

### AWS Lambda wrapper code

> This file will contain all the OpenTelemetry logic, which will enable tracing.

Reviewer 2: ```suggestion
This file contains all the OpenTelemetry logic, which enables tracing.
```

> Please save the following code as `lambda-wrapper.js`.

Reviewer 3: This involves reproducing a substantial portion of the OTel Lambda layers.  Should the documentation provide guidance on how to use those layers instead of directing the user to create their own wrapper?

Reviewer 2: Hmmm. Would there be any benefit to _not_ configuring and using the layer? I would say yes, we should show that, but I would want a good accounting of the pros and cons.

Author: I tried making the documentation as simple as possible to get the user started with tracing their Lambdas, letting them know the hows and the whys.
Using the layer is also possible, but it has to be compiled and deployed by the user.

Reviewer: Via the [AWS Distro docs](https://aws-otel.github.io/docs/getting-started/lambda/lambda-js) the description _"Users can enable and disable OpenTelemetry for their Lambda function without changing code."_ sounds like we have a "manual vs automatic instrumentation" situation here? 

Agreed with @cartermp we need to have a good understanding of the Pros/Cons of using the one vs the other, that makes it easier to decide if we should have both or only one of them here in the docs. 

If the answer is "both" for the sake of closing this PR we could start with an extra section that calls out the existence of the layer, would that work?

Reviewer 3: I think this documentation should at the least call out the existence of the layers and note that it duplicates a subset of their functionality.  It's good to have the DIY option spelled out, but I expect that many users are not going to want to maintain this themselves if an alternative is available.

Reviewer: > I think this documentation should at the least call out the existence of the layers and note that it duplicates a subset of their functionality. It's good to have the DIY option spelled out, but I expect that many users are not going to want to maintain this themselves if an alternative is available.

Makes sense. @samimusallam can you put a pointer to the layers for now. In a follow up we can maybe have both in the docs.

Author: I've added a reference to it at the beginning of the docs.

```javascript
/* lambda-wrapper.js */

const api = require("@opentelemetry/api");
const { BatchSpanProcessor } = require("@opentelemetry/sdk-trace-base");
const {
  OTLPTraceExporter,
} = require("@opentelemetry/exporter-trace-otlp-http");
const { NodeTracerProvider } = require("@opentelemetry/sdk-trace-node");
const { registerInstrumentations } = require("@opentelemetry/instrumentation");
const {
  getNodeAutoInstrumentations,
} = require("@opentelemetry/auto-instrumentations-node");

api.diag.setLogger(new api.DiagConsoleLogger(), api.DiagLogLevel.ALL);

const provider = new NodeTracerProvider();
const collectorOptions = {
  url: "<backend_url>",
};

const spanProcessor = new BatchSpanProcessor(
  new OTLPTraceExporter(collectorOptions)

> );

Reviewer 2: Hmmm. I am not sure about this. My understanding is that batches can get lost on AWS Lambda. Don't know about other serverless environments. But I'm not an expert here either.

Reviewer: Probably there needs to be a [force flush](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/sdk.md#forceflush) somewhere?

Reviewer 2: I think that can also hurt performance in general. I'd be okay with getting this out there as a first cut but we can't have things documented that drop spans or significantly screw up performance.

Maybe @Aneurysm9 has some suggestions on what to document here?

Author: The BatchSpanProcessor does flush on [shutdown](https://github.com/open-telemetry/opentelemetry-js/blob/27c4f762948c9b8d693d5ce1c2df5aaa5b0e92ee/packages/opentelemetry-sdk-trace-base/src/export/BatchSpanProcessorBase.ts#L99), so it should send all traces before the lambda shutdown.
I'm not sure how big of a performance difference it is compared to a SimpleSpanProcessor for example, but the BatchSpanProcessor is also used in the [OpenTelemetry AWS Lambda Layer](https://github.com/open-telemetry/opentelemetry-lambda/blob/3971af6d6f47e9dc138f9d45448880e3ba8ae37f/nodejs/packages/layer/src/wrapper.ts#L101)

Reviewer 3: It is generally recommended to use the `BatchSpanProcessor` everywhere but in testing where blocking export on span end can be important and tolerated.  With Lambda functions it is generally advised to call `ForceFlush` on the `(Trace|Meter)Provider` at the end of your function invocation to ensure that any telemetry is emitted before the runtime can be frozen.

There are [known issues](https://github.com/open-telemetry/opentelemetry-lambda/issues/224) with telemetry getting dropped by the Lambda layers, but they relate to extension lifecycle management issues in the collector extension layer and should not be affected by choice of span processor in the function space.

Reviewer 2: Makes sense! Yeah, in this case ensuring there's a call at the end of the function invocation to forceFlush should then suffice.

provider.addSpanProcessor(spanProcessor);
provider.register();

registerInstrumentations({
  instrumentations: [
    getNodeAutoInstrumentations({
      "@opentelemetry/instrumentation-aws-lambda": {
        disableAwsContextPropagation: true,
      },
    }),
  ],
});
```

Replace `<backend_url>` with the URL of your favorite backend to export all
traces to it. If you don't have one setup already, you can check out [Jaeger](https://www.jaegertracing.io/) or [Zipkin](https://zipkin.io/).

Note that `disableAwsContextPropagation` is set to true. The reason for this is
that the Lambda instrumentation tries to use the X-Ray context headers by
default, this results in a non-sampled context, which creates a

> `NonRecordingSpan`.

Reviewer 3: I believe this is only true if active tracing is disabled on your function.  When it is enabled then a sampled trace context will be available.

Author: True, since by default active tracing is not enabled, I've set this option to true

Reviewer 3: ```suggestion
Note that `disableAwsContextPropagation` is set to true. The reason for this is
that the Lambda instrumentation tries to use the X-Ray context headers by
default, this results in a non-sampled context unless active tracing is enabled for this function, which creates a
`NonRecordingSpan`.
```

That should be made explicit in the text.  Readers may be confused and think it is not at all possible to get a recording span or to use the AWS context propagation carrier.

Author: You're right, it should be noted. I pushed an update!

More details can be found in the instrumentation
[documentation](https://www.npmjs.com/package/@opentelemetry/instrumentation-aws-lambda).

### AWS Lambda function handler

> Now that you have a Lambda wrapper, you can create a simple handler that will
> serve as a Lambda function. Save the following code as `handler.js`.

Reviewer 2: ```suggestion
Now that you have a Lambda wrapper, create a simple handler that servers
as a Lambda function. Save the following code as `handler.js`.
```

```javascript
/* handler.js */

"use strict";

const https = require('https');

> function getRequest() {

Reviewer 2: Is `async` and `await` possible to use in lambda? Would prefer to show that instead of promises if possible

Author: I went with this as a reference since I want to wait for the response and return it
https://docs.aws.amazon.com/lambda/latest/dg/nodejs-handler.html#nodejs-handler-async

  const url = 'https://opentelemetry.io/';

  return new Promise((resolve, reject) => {
    const req = https.get(url, res => {
      resolve(res.statusCode);
    });

    req.on('error', err => {
      reject(new Error(err));
    });
  });
}

exports.handler = async (event) => {
  try {
    const result = await getRequest();
    return {
      statusCode: result,
    };
  } catch (error) {
    return {
      statusCode: 400,
      body: error.message,
    };
  }
};
```

### Deployment

> There are multiple ways of deploying, but we will be using [Serverless Framework](https://github.com/serverless/serverless)

Author: I'll link the Serverless Framework's Getting Started guide.
Other than using AWS Console or AWS CLI, I'm not familiar with other deployments methods

> There are multiple ways of deploying, but we will be using [Serverless Framework](https://github.com/serverless/serverless)

Reviewer: As long as we clarify that this is _one_ of multiple ways of doing it, I have no issue with being prescriptive in a first version of the doc. Later we still can provide options. @samimusallam besides having the links to the serverless framework maybe call out those other options without documenting them, e.g.

```suggestion
There are multiple ways of deploying your lambda function:

  * AWS Console
  * AWS CLI
  * [Serverless Framework](https://github.com/serverless/serverless)
  * ...
  
Here we will be using Serverless Framework
```

> There are multiple ways of deploying, but we will be using [Serverless Framework](https://github.com/serverless/serverless)

Author: I've added a list of other options with their respective links

for ease of use. Create a file called `serverless.yml`:

```yaml
service: lambda-otel-native
frameworkVersion: "3"
provider:
  name: aws
  runtime: nodejs14.x
  region: "<your-region>"
  environment:
    NODE_OPTIONS: --require lambda-wrapper
functions:
  lambda-otel-test:
    handler: handler.hello
```

For OpenTelemetry to work properly, `lambda-wrapper.js` must be included before
any other file: the `NODE_OPTIONS` setting ensures this.

Note if you are not using Serverless Framework to deploy your Lambda function,

> you must manually add this environment variable using the AWS Console UI.

Reviewer 2: This seems to me like we need an additional section for how to properly package and deploy the app for Lambda when not using serverles.

serve as a Lambda function. Save the following code as `handler.js`.

```javascript
/* handler.js */

"use strict";

const https = require('https');

function getRequest() {
  const url = 'https://opentelemetry.io/';

  return new Promise((resolve, reject) => {
    const req = https.get(url, res => {
      resolve(res.statusCode);
    });

    req.on('error', err => {
      reject(new Error(err));
    });
  });
}

exports.handler = async (event) => {
  try {
    const result = await getRequest();
    return {
      statusCode: result,
    };
  } catch (error) {
    return {
      statusCode: 400,
      body: error.message,
    };
  }
};
```

### Deployment

There are multiple ways of deploying, but we will be using [Serverless Framework](https://github.com/serverless/serverless)
for ease of use. Create a file called `serverless.yml`:

```yaml
service: lambda-otel-native
frameworkVersion: "3"
provider:
  name: aws
  runtime: nodejs14.x
  region: "<your-region>"
  environment:
    NODE_OPTIONS: --require lambda-wrapper
functions:
  lambda-otel-test:
    handler: handler.hello
```

For OpenTelemetry to work properly, `lambda-wrapper.js` must be included before

> any other file. That's why we have added the environment variable
> `NODE_OPTIONS:--require lambda-wrapper` which preloads the wrapper at startup.

Reviewer 4: ```suggestion
any other file: the `NODE_OPTIONS` setting ensures this.
```

Author: Looks a lot better, thanks! :)
